worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package raftbackstop
+
+import (
+	"bytes"
+
+	"github.com/hashicorp/raft"
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/pubsub"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/hashicorp/go-msgpack/codec"
+	"github.com/juju/juju/pubsub/apiserver"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+// RaftNode captures the part of the *raft.Raft API needed by the
+// backstop worker.
+type RaftNode interface {
+	State() raft.RaftState
+	GetConfiguration() raft.ConfigurationFuture
+}
+
+// This worker monitors the state of the raft cluster it's passed, and
+// if it detects that there is only one remaining API server that is
+// non-voting, will force-append a configuration message to the raft
+// log to enable it to recover. This allows us to handle the situation
+// where we had two remaining machines (a leader and a non-voting
+// follower) but the leader was removed.
+
+// Config holds the values needed by the worker.
+type Config struct {
+	Raft     RaftNode
+	LogStore raft.LogStore
+	Hub      *pubsub.StructuredHub
+	Logger   loggo.Logger
+	Tag      names.Tag
+}
+
+// Validate validates the raft worker configuration.
+func (config Config) Validate() error {
+	if config.Hub == nil {
+		return errors.NotValidf("nil Hub")
+	}
+	if config.Raft == nil {
+		return errors.NotValidf("nil Raft")
+	}
+	if config.LogStore == nil {
+		return errors.NotValidf("nil LogStore")
+	}
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	return nil
+}
+
+// NewWorker returns a worker responsible for recovering the raft
+// cluster when it's been reduced to one server that can't become
+// leader.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &Worker{
+		config:        config,
+		serverDetails: make(chan apiserver.Details),
+	}
+	// Subscribe to API server address changes.
+	unsubscribe, err := config.Hub.Subscribe(
+		apiserver.DetailsTopic,
+		w.apiserverDetailsChanged,
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "subscribing to apiserver details")
+	}
+	// Now that we're subscribed, request the current API server details.
+	req := apiserver.DetailsRequest{
+		Requester: "raft-backstop",
+		LocalOnly: true,
+	}
+	if _, err := config.Hub.Publish(apiserver.DetailsRequestTopic, req); err != nil {
+		return nil, errors.Annotate(err, "requesting current apiserver details")
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer unsubscribe()
+			return w.loop()
+		},
+	}); err != nil {
+		unsubscribe()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+type Worker struct {
```

Comment:
Does this need to be exported?

Reply:
No, I think I just called it that to avoid clobbering the worker package import. Could rename it to `backstopWorker`?

Reply:
Renamed.